### PR TITLE
Refactor slow sentencepiece tokenizers.

### PR DESCRIPTION
**Pull request by PhilipMay**
PR for #11646

## ToDo

- [x] `AlbertTokenizer`
- [x] `BarthezTokenizer`
- [x] `BertGenerationTokenizer`
- [x] `BigBirdTokenizer`
- [x] `CamembertTokenizer`
- [x] `DebertaV2Tokenizer`
- [x] `M2M100Tokenizer`
- [x] `MarianTokenizer`
- [x] `MBart50Tokenizer`
- [x] `PegasusTokenizer`
- [x] `ReformerTokenizer`
- [x] `Speech2TextTokenizer`
- [x] `T5Tokenizer`
- [x] `XLMProphetNetTokenizer`
- [x] `XLM RoBERTa`
- [x] `XLNetTokenizer`

### CODE REVIEW

**Comment by PhilipMay** [comment]
`SentencePieceProcessor.decode` is doing "the same but more than `SentencePieceProcessor.decode_pieces`. 
That is why we replace `SentencePieceProcessor.decode_pieces` with `SentencePieceProcessor.decode` in this PR.

See here:

https://github.com/google/sentencepiece/blob/6256ef243844e5848499cf519eb2a7e2755e75a1/python/src/sentencepiece/__init__.py#L307

**Comment by PhilipMay** [comment]
rebased on upstrem/master

**Comment by PhilipMay** [comment]
We need to rebase on master after PR #11737 has been merged.

**Comment by PhilipMay** [comment]
Rebased on master - CI is green again. :-)

**Comment by PhilipMay** [comment]
Rebased on master to get integration tests - see #11737

**Comment by PhilipMay** [comment]
Rebased on master

File: src/transformers/tokenization_utils.py
Comment (PhilipMay) [inline src/transformers/tokenization_utils.py:832]: Hey @LysandreJik - I would like to hear your feedback about this function.
Is it cool to refactor it into the base class? Or is it overengineered?

Thanks
Philip

Comment (LysandreJik) [reply src/transformers/tokenization_utils.py:832]: I think generally speaking we'd like to have methods that are common to *all* tokenizers in the base class - but not methods that are common to some of them only. I'd also like to keep the number of abstraction layers to a minimum, tokenizers are already quite tough to understand.

**Review by LysandreJik — COMMENTED** [review]
This is an interesting proposal! I'm not sure I understand everything, so I'm asking a few questions :)

File: src/transformers/tokenization_utils.py
Comment (LysandreJik) [inline src/transformers/tokenization_utils.py:777]: I'm not too keen on having an additional abstraction layer `PreTrainedSentencepieceTokenizer`. 

I thought the original idea was to replace instances of `"".join(tokens).replace(SPIECE_UNDERLINE, " ").strip() ` by `self._tokenizer.decode(tokens) `.

Could you explain why the changes proposed here are necessary?

Comment (PhilipMay) [reply src/transformers/tokenization_utils.py:777]: > Could you explain why the changes proposed here are necessary?

Well. While doing this refactoring I noticed a lot of duplicate code in the tokenizers implementations.
Since wet code is hard to maintain I tried to refactor it.

But if you do not like my refactoring I can just do the single change and that's it.

**Comment by PhilipMay** [comment]
> I think generally speaking we'd like to have methods that are common to all tokenizers in the base class - but not methods that are common to some of them only. I'd also like to keep the number of abstraction layers to a minimum, tokenizers are already quite tough to understand.

Yes. I also prefer a low number of abstraction layers. At the same time I like dry code. There is 100% duplicate code in the tokenizers impl. that has just been duplicated by copy & paste. IMO that should be removed by an refactoring. That is what I try to introduce here.

**Comment by LysandreJik** [comment]
The general approach of the library is to keep the number of abstractions as low as possible, and to keep implementations as separate as possible from each other, hence the high amount of copy-pasted code.

We want users to be able to experiment with single models/tokenizers without their changes impacting other models or tokenizers - and we want them to be able to understand how a model or tokenizer behaves by simply checking a single file, rather than having to hop around multiple files.

We are failing at this with tokenizers as there are already two levels of abstraction, but adding a third one isn't really the direction we want to head to :)

Does that make sense?

**Comment by PhilipMay** [comment]
> Does that make sense?

Yes. Sure. Your project, your call.
I will reverse my changes and keep it as simple as possible as discussed in the beginning.

**Comment by PhilipMay** [comment]
@LysandreJik I have redone the PR. Everything is green and the changes are as simple as planned in the issue.
This is ready for review.

Averything is tested by setting `test_sentencepiece = True` in the tokenizer test classes and by the following 
testfunction: `TokenizerTesterMixin.test_sentencepiece_tokenize_and_convert_tokens_to_string`

**Review by LysandreJik — APPROVED** [review]
Thank you @PhilipMay, LGTM!

**Pull request merged**
